### PR TITLE
Support seeking over HTTP and HTTPS

### PR DESCRIPTION
**Pull request by interpolatio**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -272,7 +272,27 @@ Since going over all (or select) keys in an S3 bucket is a very common operation
...
+  >>> from smart_open import open
+  >>> with open('s3://bucket/key.txt', 'rb', transport_params = {'version_id':'need_key_version_id'}) as fin:
+  ...   for line in fin:
+  ......   print(line)
+  b'sdsaf\r\n'
+  b'1234\r\n'
+  b'asdf\r\n'
+  b'23154ffd\r\n'
+  b's'
+
```
Comment (mpenkov) [inline README.rst]: Roll back the changes to the readme, they don't belong to this PR.

File: smart_open/tests/test_http.py [hunk 2]
```diff
@@ -107,3 +108,24 @@ def test_headers(self):
...
+            self.assertEqual(BYTES[10:20], read_bytes)
+
+    @responses.activate
+    def test_https_seek_revert(self):
+        """Did the seek revert over HTTPS work?"""
+        responses.add_callback(responses.GET, URLS, callback=request_callback)
+
+        with smart_open.open(URLS, "rb") as fin:
+            read_bytes_1 = fin.read(size=10)
+            fin.seek(-10, whence=smart_open.s3.CURRENT)
```
Comment (mpenkov) [inline smart_open/tests/test_http.py]: This is good. Add another test case with fin.seek(0) (seek to start of file) as well, because that's also a common use case.

File: smart_open/tests/test_http.py [hunk 3]
```diff
@@ -8,6 +8,7 @@
 
 BYTES = b'i tried so hard and got so far but in the end it doesn\'t even matter'
 URL = 'http://localhost'
+URLS = 'https://localhost'
```
Comment (mpenkov) [inline smart_open/tests/test_http.py]: HTTPS_URL is better

File: smart_open/tests/test_http.py [hunk 4]
```diff
@@ -107,3 +108,24 @@ def test_headers(self):
         responses.add_callback(responses.GET, URL, callback=request_callback)
         reader = smart_open.http.open(URL, 'rb', headers={'Foo': 'bar'})
         self.assertEqual(reader.headers['Foo'], 'bar')
+
+    @responses.activate
+    def test_https_seek_forward(self):
+        """Did the seek forward over HTTPS work?"""
+        responses.add_callback(responses.GET, "https://localhost", callback=request_callback)
```
Comment (mpenkov) [inline smart_open/tests/test_http.py]: Use HTTPS_URL constant, here and below

File: smart_open/tests/test_http.py [hunk 5]
```diff
@@ -107,3 +108,34 @@ def test_headers(self):
...
+        """Did the seek forward over HTTPS work?"""
+        responses.add_callback(responses.GET, HTTPS_URL, callback=request_callback)
+
+        with smart_open.open(HTTPS_URL, "rb") as fin:
+            fin.seek(10)
+            read_bytes = fin.read(size=10)
+            self.assertEqual(BYTES[10:20], read_bytes)
+
+    @responses.activate
+    def test_https_seek_revert(self):
```
Comment (mpenkov) [inline smart_open/tests/test_http.py]:
```suggestion
    def test_https_seek_reverse(self):
```

File: smart_open/tests/test_http.py [hunk 6]
```diff
@@ -107,3 +108,34 @@ def test_headers(self):
...
+        """Did the seek revert over HTTPS work?"""
```
Comment (mpenkov) [inline smart_open/tests/test_http.py]:
```suggestion
        """Did the seek in reverse over HTTPS work?"""
```

File: smart_open/tests/test_http.py (hunk 2 above)
Comment (interpolatio) [reply smart_open/tests/test_http.py]: Ok
https://github.com/interpolatio/smart_open/commit/20179e5bba9d18bc6e4dd337c7382f48b9ac8ac9

File: smart_open/tests/test_http.py [hunk 7]
```diff
@@ -107,3 +108,35 @@ def test_headers(self):
         responses.add_callback(responses.GET, URL, callback=request_callback)
         reader = smart_open.http.open(URL, 'rb', headers={'Foo': 'bar'})
         self.assertEqual(reader.headers['Foo'], 'bar')
+
+    @responses.activate
+    def test_https_seek_start(self):
+        """Did the seek start over HTTPS work?"""
```
Comment (mpenkov) [inline smart_open/tests/test_http.py]:
```suggestion
        """Does seeking to the start of the file work?"""
```

File: smart_open/tests/test_http.py [hunk 8]
```diff
@@ -107,3 +108,35 @@ def test_headers(self):
...
+
+        with smart_open.open(HTTPS_URL, "rb") as fin:
+            read_bytes_1 = fin.read(size=10)
+            fin.seek(0)
+            read_bytes_2 = fin.read(size=10)
+            self.assertEqual(read_bytes_1, read_bytes_2)
+
+    @responses.activate
+    def test_https_seek_forward(self):
+        """Did the seek forward over HTTPS work?"""
```
Comment (mpenkov) [inline smart_open/tests/test_http.py]:
```suggestion
        """Does seeking to the middle of the file work?"""
```

File: smart_open/tests/test_http.py [hunk 9]
```diff
@@ -107,3 +108,35 @@ def test_headers(self):
...
+        responses.add_callback(responses.GET, HTTPS_URL, callback=request_callback)
+
+        with smart_open.open(HTTPS_URL, "rb") as fin:
+            fin.seek(10)
+            read_bytes = fin.read(size=10)
+            self.assertEqual(BYTES[10:20], read_bytes)
+
+    @responses.activate
+    def test_https_seek_reverse(self):
+        """Did the seek in reverse over HTTPS work?"""
```
Comment (mpenkov) [inline smart_open/tests/test_http.py]:
```suggestion
        """Does seeking in reverse work?"""
```

**Comment by mpenkov** [comment]
OK, merged. Well done :fireworks:!

**Comment by mpenkov** [comment]
Fix #331
